File: swift/lpos.swift
struct Block {
    var data: String
    var previousHash: String
    var stake: Double
    var validator: String
    var hash: String

    init(data: String, previousHash: String, stake: Double, validator: String) {
        self.data = data
        self.previousHash = previousHash
        self.stake = stake
        self.validator = validator
        self.hash = calculateHash()
    }

    func calculateHash() -> String {
        let dataString = data + previousHash + String(stake) + validator
        let dataData = dataString.data(using: .utf8)!
        let hash = SHA256.hash(data: dataData)
        return hash.map { String(format: "%02x", $0) }.joined()
    }
}

class Blockchain {
    var chain: [Block]
    var validators: [(address: String, stake: Double)]
    var blockTime: Double

    init() {
        chain = [createGenesisBlock()]
        validators = []
        blockTime = 5000
    }

    func createGenesisBlock() -> Block {
        return Block(data: "Genesis Block", previousHash: "0", stake: 0, validator: "")
    }

    func getLatestBlock() -> Block {
        return chain[chain.count - 1]
    }

    func addBlock(newBlock: Block) {
        newBlock.previousHash = getLatestBlock().hash
        chain.append(newBlock)
    }

    func isChainValid() -> Bool {
        for i in 1..<chain.count {
            let currentBlock = chain[i]
            let previousBlock = chain[i - 1]

            if currentBlock.hash != currentBlock.calculateHash() {
                return false
            }

            if currentBlock.previousHash != previousBlock.hash {
                return false
            }
        }
        return true
    }

    func addValidator(validator: String, stake: Double) {
        validators.append((address: validator, stake: stake))
    }

    func selectValidator(stake: Double) -> String {
        var selectedValidator = ""
        var maxStake = 0.0

        for validator in validators {
            if validator.stake > maxStake {
                selectedValidator = validator.address
                maxStake = validator.stake
            }
        }

        return selectedValidator
    }

    func validateBlock(block: Block) -> Bool {
        return block.validator == selectValidator(stake: block.stake)
    }
}

File: js/lpos.js
class Block {
    constructor(data, previousHash, stake, validator) {
        this.data = data;
        this.previousHash = previousHash;
        this.stake = stake;
        this.validator = validator;
        this.hash = this.calculateHash();
    }

    calculateHash() {
        return CryptoJS.SHA256(this.data + this.previousHash + this.stake + this.validator).toString();
    }
}

class Blockchain {
    constructor() {
        this.chain = [this.createGenesisBlock()];
        this.validators = [];
        this.blockTime = 5000;
    }

    createGenesisBlock() {
        return new Block("Genesis Block", "0", 0, "");
    }

    getLatestBlock() {
        return this.chain[this.chain.length - 1];
    }

    addBlock(newBlock) {
        newBlock.previousHash = this.getLatestBlock().hash;
        this.chain.push(newBlock);
    }

    isChainValid() {
        for (let i = 1; i < this.chain.length; i++) {
            const currentBlock = this.chain[i];
            const previousBlock = this.chain[i - 1];

            if (currentBlock.hash !== currentBlock.calculateHash()) {
                return false;
            }

            if (currentBlock.previousHash !== previousBlock.hash) {
                return false;
            }
        }
        return true;
    }

    addValidator(validator, stake) {
        this.validators.push({
            address: validator,
            stake: stake
        });
    }

    selectValidator(stake) {
        let selectedValidator = "";
        let maxStake = 0;

        for (let i = 0; i < this.validators.length; i++) {
            if (this.validators[i].stake > maxStake) {
                selectedValidator = this.validators[i].address;
                maxStake = this.validators[i].stake;
            }
        }

        return selectedValidator;
    }

    validateBlock(block) {
        return block.validator === this.selectValidator(block.stake);
    }
}
